### PR TITLE
ARROW-11220: [Rust] Implement GROUP BY support for Boolean

### PR DESCRIPTION
**Pull request by ovr**
Introduce support in DataFushion for GROUP BY on boolean values. Boolean type in Rust implements Eq and Hash traits which allow us to use GroupByScalar.

### CODE REVIEW

**Review by andygrove — APPROVED** [review]
LGTM

File: rust/datafusion/src/physical_plan/hash_join.rs [hunk 1]
```diff
@@ -447,6 +447,11 @@ pub(crate) fn create_key(
                 // store the string value
                 vec.extend_from_slice(value.as_bytes());
             }
+            DataType::Boolean => {
+                let array = col.as_any().downcast_ref::<BooleanArray>().unwrap();
+                let x: u8 = if array.value(row) { 1 } else { 0 };
```
Comment (Dandandan) [inline rust/datafusion/src/physical_plan/hash_join.rs:452]: This could probably also use `array.value(row) as u8`?

**Review by Dandandan — APPROVED** [review]
LGTM. I added one style comment. Thanks @ovr !

**Review by alamb — COMMENTED** [review]
I suggest adding an end-to-end sql test in https://github.com/apache/arrow/blob/master/rust/datafusion/tests/sql.rs to make sure the plumbing is all hooked up correctly

**Comment by alamb** [comment]
I think this PR needs a rebase and perhaps an end-to-end test for grouping (as you did on https://github.com/apache/arrow/pull/9175)  and it will be ready to go. Thanks again @ovr !

File: rust/datafusion/src/physical_plan/hash_join.rs (hunk 1 above)
Comment (ovr) [reply rust/datafusion/src/physical_plan/hash_join.rs:460]: Thanks

**Comment by ovr** [comment]
@alamb Rebased, I added test. Thanks

**Review by alamb — APPROVED** [review]
Looks great -- thanks @ovr

**Comment by alamb** [comment]
I apologize for the delay in merging Rust PRs -- the 3.0 release is being finalized now and are planning to minimize entropy by postponing merging  changes not critical for the release until the process was complete. I hope the process is complete in the next few days. There is more [discussion](https://lists.apache.org/list.html?dev@arrow.apache.org) in the mailing list

**Comment by alamb** [comment]
Thanks again for the contribution @ovr

**Pull request closed without merging**